src/broadcast.js

```diff
@@ -0,0 +1,230 @@
+//
+// Copyright 2019 DxOS.
+//
+
+import assert from 'assert';
+import { EventEmitter } from 'events';
+import crypto from 'crypto';
+import Codec from '@dxos/codec-protobuf';
+import debug from 'debug';
+
+// eslint-disable-next-line
+import schema from './schema.json';
+import TimeLRUSet from './time-lru-set';
+
+debug.formatters.h = v => v.toString('hex').slice(0, 6);
+const log = debug('broadcast');
+
+const msgId = (seqno, from) => {
+  assert(Buffer.isBuffer(seqno));
+  assert(Buffer.isBuffer(from));
+  return `${seqno.toString('hex')}:${from.toString('hex')}`;
+};
+
+/**
+ * @typedef {Object} Middleware
+ * @property {Function} lookup Async peer lookup.
+ * @property {Function} send Defines how to send the packet builded by the broadcast.
+ * @property {Function} subscribe Defines how to subscribe to incoming packets.
+ */
+
+/**
+ * @typedef {Object} Packet
+ * @property {Buffer} seqno Id message.
+ * @property {Buffer} origin Represents the author's ID of the message. To identify a message (`msgId`) in the network you should check for the: `seqno + origin`.
+ * @property {Buffer} from Represents the current sender's ID of the message.
+ * @property {Buffer} data Represents an opaque blob of data, it can contain any data that the publisher wants to send.
+ */
+
+/**
+ * Broadcast
+ *
+ * Abstract module to send broadcast messages.
+ *
+ * @extends {EventEmitter}
+ */
+export class Broadcast extends EventEmitter {
+  /**
+   * @constructor
+   * @param {Middleware} middleware
+   * @param {Object} options
+   * @param {Buffer} options.id Defines an id for the current peer.
+   * @param {number} [options.maxAge=10000] Defines the max live time for the cache messages.
+   * @param {number} [options.maxSize=100] Defines the max size for the cache messages.
+   */
+  constructor (middleware, options = {}) {
+    assert(middleware);
```

Comment:
Should we use this always (other modules?) rather than console.assert? I like that it can happen before super().
What are the pros/cons?

Reply:
`assert` throws an error when a condition is not valid.